Source/Core/Core/AchievementManager.cpp

```diff
@@ -239,6 +239,37 @@ void AchievementManager::DoFrame()
   }
 }
 
+bool AchievementManager::CanPause()
+{
+  u32 frames_to_next_pause = 0;
+  bool can_pause = rc_client_can_pause(m_client, &frames_to_next_pause);
+  if (!can_pause)
+  {
+    OSD::AddMessage("Cannot spam pausing in hardcore mode.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    OSD::AddMessage(
+        fmt::format("Can pause in {} seconds.", static_cast<float>(frames_to_next_pause) / 60),
+        OSD::Duration::VERY_LONG, OSD::Color::RED);
+  }
+  return can_pause;
+}
+
+void AchievementManager::DoIdle()
+{
+  m_idle_thread = std::make_unique<std::thread>([]() {
+    while (Core::GetState(*AchievementManager::GetInstance().m_system) == Core::State::Paused)
+    {
+      Common::SleepCurrentThread(1000);
+      std::lock_guard lg{AchievementManager::GetInstance().GetLock()};
```

Comment:
The other thread isn't locking this lock when reassigning m_client.

Reply:
I also don't recommend using the same lock for every piece of state that needs locking, as this increases lock contention. If you need to protect `m_client`, you can have a lock for just `m_client`.